sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,481 @@
+        info!(
+            &self.log,
+            "Wrote artifact";
+            "sha256" => &sha256.to_string(),
+            "path" => final_path.as_str()
+        );
+        Ok(())
+    }
+
+    /// PUT operation (served by Sled Agent API) which takes a [`StreamingBody`]
+    pub(crate) async fn put_body(
+        &self,
+        sha256: ArtifactHash,
+        body: StreamingBody,
+    ) -> Result<(), Error> {
+        self.put(sha256, body.into_stream().map_err(Error::Body)).await
+    }
+
+    /// POST operation (served by Sled Agent API)
+    pub(crate) async fn copy_from_depot(
+        &self,
+        sha256: ArtifactHash,
+        address: SocketAddrV6,
+    ) -> Result<(), Error> {
+        static CLIENT: LazyLock<reqwest::Client> = LazyLock::new(|| {
+            reqwest::ClientBuilder::new()
+                .connect_timeout(Duration::from_secs(15))
+                .read_timeout(Duration::from_secs(15))
+                .build()
+                .unwrap()
+        });
+
+        let client = repo_depot_client::Client::new_with_client(
+            &format!("http://{address}"),
+            CLIENT.clone(),
+            self.log.new(
+                slog::o!("component" => "Repo Depot client (ArtifactStore)"),
+            ),
+        );
+        let stream = client
+            .artifact_get_by_sha256(&sha256.to_string())
+            .await
+            .map_err(|err| Error::DepotCopy { sha256, address, err })?
+            .into_inner()
+            .into_inner()
+            .map_err(|err| Error::DepotCopy {
+                sha256,
+                address,
+                err: repo_depot_client::ClientError::ResponseBodyError(err),
+            });
+        self.put(sha256, stream).await
+    }
+
+    /// DELETE operation (served by Sled Agent API)
+    pub(crate) async fn delete(
+        &self,
+        sha256: ArtifactHash,
+    ) -> Result<(), Error> {
+        let sha256 = sha256.to_string();
+        let mut any_datasets = false;
+        for dataset in self.datasets().await? {
+            any_datasets = true;
+            let path = dataset.join(&sha256);
+            match tokio::fs::remove_file(&path).await {
+                Ok(()) => {
+                    info!(
+                        &self.log,
+                        "Removed artifact";
+                        "sha256" => &sha256,
+                        "path" => path.as_str()
+                    );
+                }
+                Err(err) if err.kind() == ErrorKind::NotFound => {}
+                Err(err) => {
+                    return Err(Error::File { verb: "remove", path, err });
+                }
+            }
+        }
+        if any_datasets {
+            Ok(())
+        } else {
+            // If we're here because there aren't any update datasets, we should
+            // report Service Unavailable instead of a successful result.
+            Err(Error::NoUpdateDataset)
+        }
+    }
+
+    async fn datasets(
```

Comment:
Nit: `dataset_mountpoints` for clarity? though maybe I'm too dataset-pilled and am used to the management of them, in addition to "where they're mounted"